src/cargo/util/toml/mod.rs

```diff
@@ -773,6 +1071,179 @@ impl<'de> de::Deserialize<'de> for VecStringOrBool {
+            }
+
+            fn visit_seq<V>(self, seq: V) -> Result<Self::Value, V::Error>
+            where
+                V: de::SeqAccess<'de>,
+            {
+                let svd = de::value::SeqAccessDeserializer::new(seq);
+                T::deserialize(svd).map(MaybeWorkspace::Defined)
+            }
+
+            fn visit_map<V>(self, map: V) -> Result<Self::Value, V::Error>
+            where
+                V: de::MapAccess<'de>,
+            {
+                let mvd = de::value::MapAccessDeserializer::new(map);
+                TomlWorkspaceField::deserialize(mvd).and_then(|t| {
+                    if t.workspace {
+                        Ok(MaybeWorkspace::Workspace)
+                    } else {
+                        Err(de::Error::custom("workspace cannot be false"))
+                    }
+                })
+            }
+        }
+
+        deserializer.deserialize_any(MaybeWorkspaceVisitor {
+            marker: PhantomData,
+        })
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+enum MaybeWorkspaceBadge {
+    Workspace(TomlWorkspaceField),
+    Defined(BTreeMap<String, BTreeMap<String, String>>),
+}
+
+/// This exists only to provide a nicer error message.
+fn deserialize_workspace_badges<'de, D>(
+    deserializer: D,
+) -> Result<Option<MaybeWorkspace<BTreeMap<String, BTreeMap<String, String>>>>, D::Error>
+where
+    D: de::Deserializer<'de>,
+{
+    match Option::deserialize(deserializer) {
+        Ok(None) => Ok(None),
+        Ok(Some(MaybeWorkspaceBadge::Defined(badges))) => Ok(Some(MaybeWorkspace::Defined(badges))),
+        Ok(Some(MaybeWorkspaceBadge::Workspace(ws))) if ws.workspace => {
+            Ok(Some(MaybeWorkspace::Workspace))
+        }
+        Ok(Some(MaybeWorkspaceBadge::Workspace(_))) => {
+            Err(de::Error::custom("workspace cannot be false"))
+        }
+
+        Err(_) => Err(de::Error::custom(
+            "expected a table of badges or { workspace = true }",
+        )),
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+struct TomlWorkspaceField {
+    workspace: bool,
+}
+
+impl<T> MaybeWorkspace<T>
+where
+    T: std::fmt::Debug + Clone,
+{
+    fn from_option(value: &Option<T>) -> Option<Self> {
+        match value {
+            Some(value) => Some(Self::Defined(value.clone())),
+            None => None,
+        }
+    }
+}
+
+/// Parses an optional field, defaulting to the workspace's value.
+fn ws_default<T, F>(
+    value: Option<MaybeWorkspace<T>>,
+    workspace: Option<&TomlWorkspace>,
+    f: F,
+    label: &str,
+) -> CargoResult<Option<T>>
+where
+    T: std::fmt::Debug + Clone,
+    F: FnOnce(&TomlWorkspace) -> &Option<T>,
+{
+    match (value, workspace) {
+        (None, _) => Ok(None),
+        (Some(MaybeWorkspace::Defined(value)), _) => Ok(Some(value)),
+        (Some(MaybeWorkspace::Workspace), Some(ws)) => f(ws)
+            .clone()
+            .ok_or_else(|| {
```

Comment:
How come this is an error? If a value isn't defined in a package or in a workspace, then the final value is `None`, right?

Reply:
Ah I see nah that makse sense, thanks!